File: Array/Binary Search/binary_search_version3.py
from typing import List


class Solution:
    def binary_search(self, nums: List[int], target: int) -> int:
        """
        The methods binary search version3 that [left, right]
        :param nums: integer list
        :param target: the target number
        :return: -1 if no result, otherwise
            the target number index in the list
        """
        # Consider the list is none or length is 0
        if len(nums) == 0 or nums is None:
            return -1

        # Initialize the left index, right index
        # The right index must be list length minus 1 dut to [left, right]
        left, right = 0, len(nums) - 1

        # If target outside list range, then returns -1
        if target < nums[left] or target > nums[right]:
            return -1

        # Use while loop to do binary search,
        # notice that while loop will stop if left equals to right
        while left + 1 < right:
            middle = left + (right - left) // 2
            # Consider three situations
            if nums[middle] < target:
                left = middle
            elif nums[middle] > target:
                right = middle
            else:
                return middle

        # The three situations when
        # left and right are adjacent
        if target == nums[left]:
            return left
        elif target == nums[right]:
            return right
        else:
            return -1


def main():
    question1 = Solution()
    solution1 = question1.binary_search([1, 3, 4, 6, 7], 4)
    print(solution1)

    question2 = Solution()
    solution2 = question2.binary_search([1, 3, 4, 6, 7], 10)
    print(solution2)


if __name__ == "__main__":
    main()

File: Array/Binary Search/binarySearchVersion3.java
package com.algorithm.array;
import org.junit.Test;

/**
 * TODO: Binary Search Version3
 *
 * @author Yaozheng Wang
 * @date 2021-12-21
 */
public class binarySearchVersion3 {
    public int findPosition (int[] numbs, int target) {
        // Consider the list is null or length is 0
        if (numbs == null || numbs.length == 0) {
            return -1;
        }

        // Initialize the left index and right index
        // to be length because of [left, right)
        int left = 0, right = numbs.length;

        // Use while loop to consider three situations
        // The left and right index cannot be the same due to [left, right)
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (numbs[mid] == target) {
                return mid;
            } else if (numbs[mid] < target) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        // No match number
        return -1;
    }

    @Test
    public static void main(String[] args) {
        binarySearchVersion3 question = new binarySearchVersion3();
        int[] numbs = new int[]{1, 2, 3, 4, 5};
        System.out.println(question.findPosition(numbs, 3));
        System.out.println(question.findPosition(numbs, 9));
    }
}
